include/lbann/weights/weights_proxy.hpp

```diff
@@ -0,0 +1,170 @@
+ */
+template <typename TensorDataType>
+class WeightsProxy
+{
+  using DataTypeWeights = data_type_weights<TensorDataType>;
+  using ValuesType = El::AbstractDistMatrix<TensorDataType>;
+  using ValuesPtrType = std::unique_ptr<ValuesType>;
+public:
+
+  /** @brief Construct an empty proxy. */
+  WeightsProxy() = default;
+
+  /** @brief Construct a proxy given the master object.
+   *
+   *  @param w Master weights object, which must have a valid storage
+   *           matrix initialized internally.
+   */
+  WeightsProxy(weights const& w)
+    : master_weights_{&w},
+      values_{setup_values_(w)}
+  {}
+
+  /** @brief Copy a WeightsProxy object.
+   *
+   *  Creates a new WeightsProxy to the same object.
+   */
+  WeightsProxy(WeightsProxy const& other)
+    : WeightsProxy(other.master_weights_)
+  {}
+
+  /** @brief Move a WeightsProxy object. */
+  WeightsProxy(WeightsProxy&& other)
+    : master_weights_{other.master_weights_},
+      values_{std::move(other.values_)}
+  {}
+
+  WeightsProxy& operator=(WeightsProxy const& other)
+  {
+    this->setup(other.master_weights_);
+    return *this;
+  }
+
+  WeightsProxy& operator=(WeightsProxy&& other)
+  {
+    WeightsProxy<TensorDataType> tmp(std::move(other));
+    this->swap(tmp);
+    return *this;
+  }
+
+  /** @brief Check if the proxy is referencing a weights object. */
+  bool empty() const noexcept { return values_ == nullptr; }
+
+  /** @brief Provide setup function for delayed construction.
+   *
+   *  This overwrites any existing data.
+   *
+   *  @param w The weights object to be proxied.
+   */
+  void setup(weights const& w)
+  {
+    auto vals = setup_values_(w);
+    master_weights_ = &w;
+    std::swap(vals, values_);
+  }
+
+  /** @brief Swap contents with another WeightsProxy object. */
+  void swap(WeightsProxy<TensorDataType>& other)
+  {
+    std::swap(master_weights_, other.master_weights_);
+    std::swap(values_, other.values_);
+  }
+
+  /** @brief Synchronize the held values with the master set. */
+  void synchronize_with_master()
+  {
+    if (!values_->Viewing())
+      El::Copy(master_weights_->get_values(), *values_);
+  }
+
+  /** @brief Access the values. */
+  ValuesType const& values() const noexcept { return *values_; }
+
+  /** @brief Access the master weights object directly. */
+  weights const& master_weights() const noexcept { return *master_weights_; }
+
+private:
+  /** @brief Establish the view of the master data. */
+  ValuesPtrType setup_values_(
+    data_type_weights<TensorDataType> const& dtw) const
+  {
+    auto const& vals = dtw.get_values();
+    ValuesPtrType ret(vals.Construct(vals.Grid(), vals.Root()));
+    El::LockedView(*ret, vals);
+    return ret;
+  }
+
+  /** @brief Establish the target matrix storage. */
+  ValuesPtrType setup_values_(weights const& w) const
+  {
+    if (auto dtw = dynamic_cast<data_type_weights<TensorDataType> const*>(&w))
+      return setup_values_(*dtw);
```

Comment:
Do we handle the case where weights have different devices or data layouts? A better way is to instantiate and align `values_` first, and then decide whether to view or align by comparing the `DistData` s. It also doesn't require any dynamic casting.